flink-tests/src/test/java/org/apache/flink/test/checkpointing/ChangelogPeriodicMaterializationITCase.java

```diff
@@ -0,0 +1,206 @@
+                                                    currentCheckpointNum
+                                                            .get()
+                                                            .compareAndSet(
+                                                                    0,
+                                                                    completedCheckpointNum.get());
+                                                    return false;
+                                                }
+                                                return true;
+                                            });
+
+                                    throwArtificialFailure();
+                                } else if (getRuntimeContext().getAttemptNumber() == 1
+                                        && currentIndex == TOTAL_ELEMENTS / 2) {
+                                    waitWhile(
+                                            () -> {
+                                                if (completedCheckpointNum.get()
+                                                        <= currentCheckpointNum.get().get()) {
+                                                    return true;
+                                                }
+                                                List<Long> allMaterializationId =
+                                                        getAllMaterializationId(
+                                                                checkpointFile, jobID);
+                                                Collections.sort(allMaterializationId);
+                                                return allMaterializationId.isEmpty()
+                                                        || !currentMaterializationId
+                                                                .get()
+                                                                .equals(allMaterializationId);
+                                            });
+                                    throwArtificialFailure();
+                                }
+                            }
+                        },
+                        jobID));
+    }
+
+    @Test
+    public void testFailedMaterialization() throws Exception {
+        File checkpointFile = TEMPORARY_FOLDER.newFolder();
+        JobID jobID = generateJobID();
+        SharedReference<AtomicBoolean> hasTriggerCheckpoint =
+                sharedObjects.add(new AtomicBoolean());
+        SharedReference<List<Long>> currentMaterializationId = sharedObjects.add(new ArrayList<>());
+        StreamExecutionEnvironment env =
+                getEnv(
+                        new DelegatedStateBackendWrapper(
+                                delegatedStateBackend,
+                                snapshotResultFuture -> {
+                                    if (hasTriggerCheckpoint.get().compareAndSet(false, true)) {
+                                        throw new RuntimeException();
```

Comment:
Rename `hasTriggerCheckpoint` to `hasFailed`?

Reply:
Is it possible that failed is one subtask and materialized another?
How about setting parallelism to 1 explicitly in `getEnv`?